src/Microsoft.Identity.Web/Resource/OpenIdConnectMiddlewareDiagnostics.cs

```diff
@@ -10,206 +10,206 @@
+
+		/// <summary>
+		/// Invoked when a request is received on the RemoteSignOutPath.
+		/// </summary>
+		private Func<RemoteSignOutContext, Task> s_onRemoteSignOut = null!;
+
+		/// <summary>
+		/// Invoked before redirecting to the identity provider to sign out.
+		/// </summary>
+		private Func<RedirectContext, Task> s_onRedirectToIdentityProviderForSignOut = null!;
+
+		/// <summary>
+		/// Invoked before redirecting to the Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri
+		/// at the end of a remote sign-out flow.
+		/// </summary>
+		private Func<RemoteSignOutContext, Task> s_onSignedOutCallbackRedirect = null!;
+
+		/// <summary>
+		/// Subscribes to all the OpenIdConnect events, to help debugging, while
+		/// preserving the previous handlers (which are called).
+		/// </summary>
+		/// <param name="events">Events to subscribe to.</param>
+		public void Subscribe(OpenIdConnectEvents events)
+		{
+			events ??= new OpenIdConnectEvents();
+
+			s_onRedirectToIdentityProvider = events.OnRedirectToIdentityProvider;
+			events.OnRedirectToIdentityProvider = OnRedirectToIdentityProviderAsync;
+
+			s_onMessageReceived = events.OnMessageReceived;
+			events.OnMessageReceived = OnMessageReceivedAsync;
+
+			s_onAuthorizationCodeReceived = events.OnAuthorizationCodeReceived;
+			events.OnAuthorizationCodeReceived = OnAuthorizationCodeReceivedAsync;
+
+			s_onTokenResponseReceived = events.OnTokenResponseReceived;
+			events.OnTokenResponseReceived = OnTokenResponseReceivedAsync;
+
+			s_onTokenValidated = events.OnTokenValidated;
+			events.OnTokenValidated = OnTokenValidatedAsync;
+
+			s_onUserInformationReceived = events.OnUserInformationReceived;
+			events.OnUserInformationReceived = OnUserInformationReceivedAsync;
+
+			s_onAuthenticationFailed = events.OnAuthenticationFailed;
+			events.OnAuthenticationFailed = OnAuthenticationFailedAsync;
+
+			s_onRemoteSignOut = events.OnRemoteSignOut;
+			events.OnRemoteSignOut = OnRemoteSignOutAsync;
+
+			s_onRedirectToIdentityProviderForSignOut = events.OnRedirectToIdentityProviderForSignOut;
+			events.OnRedirectToIdentityProviderForSignOut = OnRedirectToIdentityProviderForSignOutAsync;
+
+			s_onSignedOutCallbackRedirect = events.OnSignedOutCallbackRedirect;
+			events.OnSignedOutCallbackRedirect = OnSignedOutCallbackRedirectAsync;
+		}
+
+		private async Task OnRedirectToIdentityProviderAsync(RedirectContext context)
+		{
+			_logger.LogDebug(string.Format(CultureInfo.InvariantCulture, LogMessages.MethodBegin, nameof(OnRedirectToIdentityProviderAsync)));
+
+			await s_onRedirectToIdentityProvider(context).ConfigureAwait(false);
+
+			_logger.LogDebug("   Sending OpenIdConnect message:");
+			DisplayProtocolMessage(context.ProtocolMessage);
+			_logger.LogDebug(string.Format(CultureInfo.InvariantCulture, LogMessages.MethodEnd, nameof(OnRedirectToIdentityProviderAsync)));
+		}
+
+		private void DisplayProtocolMessage(OpenIdConnectMessage message)
```

Comment:
Can this please be ifdef'd out of existence? The reflection used in this method makes it very slow.

Reply:
developers have to set it explicitly, and i doubt they would use it in release mode. we can make a comment in the wiki, and we could add an if/def too...sometimes we might need the diagnostics in release mode.